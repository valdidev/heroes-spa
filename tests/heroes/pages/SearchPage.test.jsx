import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}));

describe('Pruebas en <SearchPage />', () => {

    beforeEach( () => jest.clearAllMocks() );

    test('Debe mostrar correctamente con valores por defecto', () => {
       const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();

        screen.debug()
    });

    test('Debe mostrar a Batman y el valor en el queryString', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']} >
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('batman');
        // screen.debug()
        const img = screen.getByRole('img');
        expect( img.src ).toContain('assets/heroes/dc-batman.jpg');
        
        const typeAlert = screen.getByLabelText('type-alert');
        expect( typeAlert.style.display ).toBe('none');
    });

    test('Debe mostrar error si no se encuentra el hero (batman123)', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        )
        // screen.debug()
        const noHeroAlert = screen.getByLabelText('no-hero-alert');
        expect( noHeroAlert.style.display ).not.toBe('none');
    });
    test('Debe llamar el navigate a la pantalla nueva', () => {
        render(
        <MemoryRouter initialEntries={['/search']}>
            <SearchPage />
        </MemoryRouter>
        )

        const input = screen.getByRole('textbox');
        fireEvent.change( input, {target: {name: 'searchText', value: 'superman'}} )

        const form = screen.getByRole('form');
        fireEvent.submit( form );

        expect( mockedUseNavigate ).toHaveBeenCalledWith('?q=superman');

    });
});