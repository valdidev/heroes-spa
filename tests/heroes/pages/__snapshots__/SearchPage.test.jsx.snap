// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> Debe mostrar correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero..."
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="type-alert"
        class="alert alert-primary animate__animated animate__flash"
      >
        Type a hero
      </div>
      <div
        aria-label="no-hero-alert"
        class="alert alert-danger animate__animated animate__bounce"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
